### PR TITLE
Release ArrayList class

### PR DESCRIPTION
**Pull request by u3IOm4uk**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,104 @@
...
+    private int capacity = DEFAULT_SIZE;
+    private int size;
+    private T[] arrayData = (T[]) new Object[DEFAULT_SIZE];
+
     @Override
     public void add(T value) {
-
+        if (size == capacity) {
+            this.grow();
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:15]: https://stackoverflow.com/a/2411283
let's only use `this` where it's necessary

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,104 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_SIZE = 10;
+    private int capacity = DEFAULT_SIZE;
+    private int size;
+    private T[] arrayData = (T[]) new Object[DEFAULT_SIZE];
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:9]: initialize arrayData in constructor

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,104 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_SIZE = 10;
+    private int capacity = DEFAULT_SIZE;
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:7]: useless field, please remove it

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,104 @@
...
+        }
+        System.arraycopy(arrayData, index, arrayData, index + 1, size - index);
+        arrayData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        int size = list.size();
+        for (int i = 0; i < size; i++) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:36]:
```suggestion
        for (int i = 0; i < list.size(); i++) {
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,104 @@
...
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:77]:
```suggestion
```

File: src/main/java/core/basesyntax/NoSuchElementException.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax;
+
+public class NoSuchElementException extends RuntimeException {
+    public NoSuchElementException(String message) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/NoSuchElementException.java:4]: why not use existing nosuchelement

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,48 +1,104 @@
...
+        return size == 0;
+    }
+
+    private void grow() {
+        capacity += capacity / 2;
+        T[] newArrayData = (T[]) new Object[capacity];
+        for (int i = 0; i < arrayData.length; i++) {
+            newArrayData[i] = arrayData[i];
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:86]: image

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,48 +1,104 @@
...
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of size " + size);
+        }
+        if (size == capacity) {
+            this.grow();
+        }
```
Comment (okuzan) [inline src/main/java/core/basesyntax/ArrayList.java:27]: move `if` inside of `grow`

**Review by okuzan — CHANGES_REQUESTED** [review]
Good job! Some improvements suggested

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,48 +1,104 @@
...
+    public ArrayList() {
+        arrayData = (T[]) new Object[DEFAULT_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == arrayData.length) {
+            this.grow();
+        }
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/ArrayList.java:18]: It is better to move this check to the method
```suggestion
        grow();
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,48 +1,104 @@
...
+        arrayData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of size " + size);
+        }
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/ArrayList.java:27]: It is better to move this check to the private method

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,48 +1,104 @@
...
+        if (size == arrayData.length) {
+            grow();
+        }
```
Comment (Ivan95kos) [inline src/main/java/core/basesyntax/ArrayList.java:30]:
```suggestion
        grow();
```

**Review by Ivan95kos — APPROVED** [review]
Great job with minor comments

**Pull request closed without merging**